### PR TITLE
[WebApi test] Disallow setting extension attributes as data array

### LINKED ISSUE #26682: Magento 2.3.4 Quote Address Extension Attribute Issue
### Preconditions (*)

1. Using Magento 2.3.4
2. USing a custom extension attribute on `Magento\Quote\Api\Data\AddressInterface` Example module here: https://github.com/NathMorgan/module-test-module

### Steps to reproduce (*)
1. Sending a post request to `estimate-shipping-methods` containing a `extension_attributes`

For example a post request to `/rest/default/V1/guest-carts/{cartId}/estimate-shipping-methods` with the following example data `{"address":{"street":["Example High Street"],"city":"Example City","region_id":null,"region":null,"country_id":"GB","postcode":"TES 01","company":"Example Company","telephone":"0000 00000","fax":null,"extension_attributes":{"test_attribute":"2"}}}`

### Expected result (*)
1. Above action completed successfully

### Actual result (*)
1. 500 Error with the message `Uncaught Error: Call to a member function setDiscounts() on array in /var/www/html/vendor/magento/module-sales-rule/Model/Quote/Discount.php: 117`

This issue seems to be introduced with this commit: https://github.com/magento/magento2/commit/6900c38941850706d9bff3aae76a428afae1343c

This seems to effect any extension attribute set on `Magento\Quote\Api\Data\AddressInterface` that is not a object

### PR DESCRIPTION
**Pull request by amenk**
Fixes #26682
While working on fix for the issue #26682 - fix already was delivered in https://github.com/magento/magento2/commit/fcd1b5808625364e09dc6aea2d9eb8c0ee209d25#diff-e2b9d1a5f4b9f3503a2f8dd072bc2d8aR360, but it contains only unit tests, while this PR contained also web api tests, that better fits for such cases.

We decided to revert all changes related to the fix and deliver only tests as part of this PR.

### CODE REVIEW

**Comment by lenaorobei** [comment]
@amenk I assume we can close this PR?

**Comment by amenk** [comment]
I have to closer check the results. Actually it seems to be a valid solution to the problem in the linked issue.

**Comment by maximbaibakov** [comment]
Looks like valid solution. 
Can we get this merged into 2.3.x as well?

**Comment by amenk** [comment]
I have no clue why functional tests fail for B2B and EE -> rerunning

**Comment by amenk** [comment]
Functional tests EE is green now without any change.
Functional Tests B2B is still red, but https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/27338/73c167e6c2bcafd4451bd73a67d6ab3f/Functional/allure-report-b2b/index.html# does not indicate any errors, so I am rerunning Functional B2B once again.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @amenk ,
Your changes looks good to me. Could you cover them with any type of tests in order to prevent appearing this issue in future?
I think API test there is the most suitable.

Note: failing functional tests seems to me as false positive, no need to fix them

**Comment by amenk** [comment]
@ihor-sviziev 
Thanks for your feedback.

I might not have the time in the near future to cover this with tests, I am happy if someone can take over.

**Comment by engcom-Kilo** [comment]
@ihor-sviziev 
Exist api-functional test 
dev/tests/api-functional/testsuite/Magento/Quote/Api/GuestShipmentEstimationTest.php::testEstimateByExtendedAddress

can be related to this Pull Request

**Comment by ihor-sviziev** [comment]
@engcom-Kilo good test example! Thank you.
@amenk maybe you could copy just duplicate that test case and a bit adjust it? I think it shouldn’t be really hard

**Comment by amenk** [comment]
@engcom-Kilo thank you
@ihor-sviziev woudn't that mean I have to do the full extension attribute setup during the test?

**Comment by ihor-sviziev** [comment]
@amenk as I understood - issue appears when you have extension attribute in the request, not in the model they you loaded. In this case it seems to me that declaring attribute is not mandatory. Let me know if I’m wrong

**Comment by amenk** [comment]
unfortunately not,

```


            $data = [
                AddressInterface::KEY_ID => (int)$address->getId(),
                AddressInterface::KEY_REGION => $address->getRegion(),
                AddressInterface::KEY_REGION_ID => $address->getRegionId(),
                AddressInterface::KEY_REGION_CODE => $address->getRegionCode(),
                AddressInterface::KEY_COUNTRY_ID => $address->getCountryId(),
                AddressInterface::KEY_STREET => $address->getStreet(),
                AddressInterface::KEY_COMPANY => $address->getCompany(),
                AddressInterface::KEY_TELEPHONE => $address->getTelephone(),
                AddressInterface::KEY_POSTCODE => $address->getPostcode(),
                AddressInterface::KEY_CITY => $address->getCity(),
                AddressInterface::KEY_FIRSTNAME => $address->getFirstname(),
                AddressInterface::KEY_LASTNAME => $address->getLastname(),
                AddressInterface::KEY_CUSTOMER_ID => $address->getCustomerId(),
                AddressInterface::KEY_EMAIL => $address->getEmail(),
                AddressInterface::SAME_AS_BILLING => $address->getSameAsBilling(),
                AddressInterface::CUSTOMER_ADDRESS_ID => $address->getCustomerAddressId(),
                AddressInterface::SAVE_IN_ADDRESS_BOOK => $address->getSaveInAddressBook(),

                'custom_attributes' => [
                    [
                    'attribute_code' =>  'company2',
                    'value' => 'ACME Division',
                    ],
                    [
                    'attribute_code' =>  'company3',
                    'value' => 'ACME Subdivision',
                    ]
                ]
            ];
```

Test does not fail, even my fix is not in place

**Comment by amenk** [comment]
Also we better would make the test for a registered user...

**Comment by amenk** [comment]
@ihor-sviziev There is a test module supplied in https://github.com/NathMorgan/module-test-module I added it in my last commit https://github.com/magento/magento2/pull/27338/commits/bc18a8fd5141a2748b4938e2468413ba7a3625ba
but my testcase is still not failing if I revert the original fix.
How are those testmodules loaded? Maybe it's not registered?

**Comment by ihor-sviziev** [comment]
@engcom-Kilo, could you help, is there any way do declare extension attribute for api tests? Maybe there some examples ?

**Comment by engcom-Kilo** [comment]
Sample for set Extension Attributes
[dev/tests/api-functional/_files/Magento/TestModuleDefaultHydrator/Model/ResourceModel/ReadHandler.php](https://github.com/magento/magento2/blob/9df4250c612b04218c11f5723d526f9a13ee7db0/dev/tests/api-functional/_files/Magento/TestModuleDefaultHydrator/Model/ResourceModel/ReadHandler.php#L91)

Seems need set Extension Attributes in fixture file

**Comment by amenk** [comment]
@engcom-Kilo I don't get it ... I tried loading a test module in my latest version of the pull request, but it did not work...

**Comment by engcom-Kilo** [comment]
> @engcom-Kilo I don't get it ... I tried loading a test module in my latest version of the pull request, but it did not work...

I will try fix test

**Comment by engcom-Kilo** [comment]
@amenk 
Please grant permission to write for iMi-digital:fix/26682-no-overwriting-extension-attributes

Thank

**Comment by amenk** [comment]
@engcom-Kilo Permssion granted, thanks :)

**Comment by engcom-Kilo** [comment]
Added API test for Extension Attribute, 
used Discounts - this Extension Attribute by default.

**Comment by amenk** [comment]
Okay, cool, very much appreciated, @engcom-Kilo - was the test failing without my patch in place?

**Comment by engcom-Kilo** [comment]
Added fix in app/code/Magento/SalesRule/Model/Quote/Discount.php
Where setted Extension Attribute, and called errors.

Issue in Extension Attributes,
in some cases returned as Array, and not available methods like setDiscounts().

Added checking, if exist Extension Attributes. And merge Extension Attributes.

**Comment by engcom-Kilo** [comment]
> Okay, cool, very much appreciated, @engcom-Kilo - was the test failing without my patch in place?

Yes,
your patch resolved this bug.

But seems not need full remove Extension Attributes.

Because Extension Attributes can be used in other modules (custom modules). 
The Extension Attributes can use like a storage and use in Observer for example.

File: app/code/Magento/SalesRule/Model/Quote/Discount.php [hunk 1]
```diff
@@ -119,7 +130,20 @@ public function collect(
 
         $address->setDiscountDescription([]);
         $items = $this->calculator->sortItemsByPriority($items, $address);
-        $address->getExtensionAttributes()->setDiscounts([]);
+
+        if (!is_object($address->getExtensionAttributes())) {
```
Comment (ihor-sviziev) [inline app/code/Magento/SalesRule/Model/Quote/Discount.php]: Hi,
Looks like this is collect totals and it shouldn’t change type of extension attributes, but getExtensionAttrubutes should return always object. I think shipping address object should come just with correct extension attributes, so place that prepares incorrect object should be adjusted instead
What do you think?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/SalesRule/Model/Quote/Discount.php (hunk 1 above)
Comment (engcom-Kilo) [reply app/code/Magento/SalesRule/Model/Quote/Discount.php]: > getExtensionAttrubutes should return always object

yes, true

Comment (engcom-Kilo) [reply app/code/Magento/SalesRule/Model/Quote/Discount.php]: Seems problem in private method **extractAddressData**
And solution by @amenk working.

I reverted own changes in Discount class.

**Comment by ihor-sviziev** [comment]
Hi @engcom-Kilo ,
Please fix static test, everything else looks good

**Comment by amenk** [comment]
@ihor-sviziev @engcom-Kilo why is database compare failing?  I guess functional b2b can be ignored.

**Comment by engcom-Kilo** [comment]
> @ihor-sviziev @engcom-Kilo why is database compare failing? I guess functional b2b can be ignored.

Yes,
these failed test not related to this Pull Request

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
@engcom-Kilo or @amenk  please fix static test, everything else looks good.

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: dev/tests/api-functional/testsuite/Magento/Quote/Api/GuestShipmentEstimationWithExtensionAttributesTest.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    const SERVICE_VERSION = 'V1';
+    const SERVICE_NAME = 'quoteGuestShipmentEstimationV1';
+    const RESOURCE_PATH = '/V1/guest-carts/';
+
+    /**
+     * @var ObjectManager
+     */
+    private $objectManager;
+
+    protected function setUp()
```
Comment (ihor-sviziev) [inline dev/tests/api-functional/testsuite/Magento/Quote/Api/GuestShipmentEstimationWithExtensionAttributesTest.php]:
```suggestion
    protected function setUp(): void
```

File: dev/tests/api-functional/testsuite/Magento/Quote/Api/GuestShipmentEstimationWithExtensionAttributesTest.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        $this->objectManager = \Magento\TestFramework\Helper\Bootstrap::getObjectManager();
+    }
+
+    /**
+     * @magentoAppIsolation enabled
+     * @magentoDbIsolation disabled
+     * @magentoApiDataFixture Magento/SalesRule/_files/cart_rule_free_shipping.php
+     * @magentoApiDataFixture Magento/Sales/_files/quote.php
+     */
+    public function testEstimateByExtendedAddress()
```
Comment (ihor-sviziev) [inline dev/tests/api-functional/testsuite/Magento/Quote/Api/GuestShipmentEstimationWithExtensionAttributesTest.php]:
```suggestion
    public function testEstimateByExtendedAddress(): void
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: dev/tests/api-functional/testsuite/Magento/Quote/Api/GuestShipmentEstimationWithExtensionAttributesTest.php (hunk 2 above)
Comment (engcom-Kilo) [reply dev/tests/api-functional/testsuite/Magento/Quote/Api/GuestShipmentEstimationWithExtensionAttributesTest.php]: Updated

File: dev/tests/api-functional/testsuite/Magento/Quote/Api/GuestShipmentEstimationWithExtensionAttributesTest.php (hunk 3 above)
Comment (engcom-Kilo) [reply dev/tests/api-functional/testsuite/Magento/Quote/Api/GuestShipmentEstimationWithExtensionAttributesTest.php]: Updated

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7563](https://jira.corp.magento.com/browse/ENGCOM-7563) has been created to process this Pull Request

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @amenk 
Unfortunately, we are not able to reproduce this [issue](https://github.com/magento/magento2/issues/26682) on a fresh 2.4-develop instance. 
It looks like this problem was fixed by the [commit](https://github.com/magento/magento2/commit/fcd1b5808625364e09dc6aea2d9eb8c0ee209d25).

**Preconditions:**
1. Magento 2.4-develop;
2. Using a custom extension attribute on `Magento\Quote\Api\Data\AddressInterface` Example module here: https://github.com/NathMorgan/module-test-module

**Manual testing scenario:**

1. Sending a post request to `estimate-shipping-methods` containing an `extension_attributes`
`{
"address":
{"street":["Example High Street"],
"city":"Example City",
"region_id":null,"region":null,
"country_id":"GB",
"postcode":"TES 01",
"company":"Example Company",
"telephone":"0000 00000",
"fax":null,
"extension_attributes":
{"test_attribute":"2"}
}
}`

**Actual Result:** :heavy_check_mark:  Response: **200 OK**

Screenshot from 2020-05-22 11-36-30
Screenshot from 2020-05-22 11-36-40

@amenk Could you take a look?

Thanks!

**Comment by amenk** [comment]
Thanks for testing.
@engcom-Alfa please confirm that you did install the test module for testing.
If the bug is fixed now, I suggest we include the tests written by @engcom-Kilo so the bug stays fixed in future releases.

**Comment by engcom-Alfa** [comment]
> Thanks for testing.
> @engcom-Alfa please confirm that you did install the test module for testing.
> If the bug is fixed now, I suggest we include the tests written by @engcom-Kilo so the bug stays fixed in future releases.

Yes, we tested with pre-installed [test module](https://github.com/NathMorgan/module-test-module)

**Comment by ihor-sviziev** [comment]
Hi @engcom-Alfa @amenk,
Looks like this issue was fixed in 
https://github.com/magento/magento2/commit/fcd1b5808625364e09dc6aea2d9eb8c0ee209d25#diff-e2b9d1a5f4b9f3503a2f8dd072bc2d8aR360

But I see it's covered only with unit test, so would be great to keep api functional test and revert all changes in code. What do you think?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
As we discussed with @engcom-Kilo in Slack - it's better to revert changes in code (as it was already fixed in 2.4-develop branch) and leave only changes in tests in order to prevent appearing this issue again in future.

@amenk or @engcom-Kilo could you revert changes in the app/code/Magento/Quote/Model/ShippingMethodManagement.php file ?

**Comment by engcom-Kilo** [comment]
Revert changes in app/code/Magento/Quote/Model/ShippingMethodManagement.php .
Also has been remove unused files.

**Comment by engcom-Kilo** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@engcom-Kilo how come earlier these files were needed and now - no?

**Comment by engcom-Kilo** [comment]
> @engcom-Kilo how come earlier these files were needed and now - no?

Checked again, 
and see that these files not working for test.

These sample files only for app/code/ .
Unused files.

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests looks not related to changes form this PR.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7563](https://jira.corp.magento.com/browse/ENGCOM-7563) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
QA not applicable

**Pull request merged**